### PR TITLE
ShaderChunk: Modulate transmission.a with transmissionFactor

### PR DESCRIPTION
**Pull request by mrdoob**
Related issue: #22425

**Description**

| Before | After |
| --- | --- |
|  Screen Shot 2021-08-31 at 7 49 48 PM | Screen Shot 2021-08-31 at 7 50 01 PM |

/cc @elalish

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/transmission_fragment.glsl.js [hunk 1]
```diff
@@ -27,6 +27,6 @@ export default /* glsl */`
 		attenuationTint, attenuationDistance );
 
 	totalDiffuse = mix( totalDiffuse, transmission.rgb, transmissionFactor );
-	transmissionAlpha = transmission.a;
+	transmissionAlpha = mix( transmissionAlpha, transmission.a, transmissionFactor );
```
Comment (elalish) [inline src/renderers/shaders/ShaderChunk/transmission_fragment.glsl.js:30]: I have no idea what's happening here, but I love the result!

**Review by elalish — APPROVED** [review]

**Comment by mrdoob** [comment]
By the way, this is how it would look like without [the `+ 0.1` hack](https://github.com/mrdoob/three.js/blob/4bf361f2b94057471fee417fd8f3e36de45b777c/src/renderers/shaders/ShaderChunk/output_fragment.glsl.js#L8):

Screen Shot 2021-08-31 at 8 11 39 PM

With the hack:

Screen Shot 2021-08-31 at 8 14 30 PM

@elalish what do you think?

**Comment by elalish** [comment]
The hack looks like it's definitely doing the right thing. The color seems pretty faint; are you sure it should be so low at 0.1? What happens adversely when it's raised?

**Comment by mrdoob** [comment]
May try to modulate with the reflection too...

**Comment by elalish** [comment]
Does it have an effect when the canvas is not set to use alpha? The idea was that it would be constrained only to changing the transparent canvas situation. 

You're absolutely right that there's no physical way to do this, but I think it's still important to make a best effort: https://github.com/mrdoob/three.js/pull/22425#issuecomment-905867432

**Comment by WestLangley** [comment]
>You're absolutely right that there's no physical way to do this

Actually, I was not concerned about being physically correct. I was concerned about being remotely correct. :-)

**Comment by mrdoob** [comment]
Unfortunately it's not possible to take a screenshot of the css background. So we are trying to get something that looks decent.

But I think this code should only affect cases where `alpha: true` is passed to `WebGLRenderer`. Are you saying this is affecting the `alpha: false` case too?

**Comment by WestLangley** [comment]
Sorry, I can not comprehend this code, so I do not know what the hacks are doing.

```glsl
export default /* glsl */`
#ifdef OPAQUE
diffuseColor.a = 1.0;
#endif

// https://github.com/mrdoob/three.js/pull/22425
#ifdef USE_TRANSMISSION
diffuseColor.a *= transmissionAlpha + 0.1;
#endif

gl_FragColor = vec4( outgoingLight, diffuseColor.a );
`;
```

I think the spheres in the test case all have `opaque` true. What is the blending function being used ? No blending? Normal blending? As far as I know, we have not agreed on what the alpha type means for three.js.

And `diffuseColor.a` can be `1.1`?  What does the compositor do with a value greater than 1? Is it clamped?

I really do not want to have to be concerned about these issues while I am trying to fix the remaining deficiencies in the shaders.

**Comment by mrdoob** [comment]
You can ignore this

```glsl
#ifdef OPAQUE
diffuseColor.a = 1.0;
#endif
```

I'm thinking about reverting the `material.format` change and setting `texture.format = THREE.RGBFormat` in `GLTFLoader` instead.

This is my understanding:
 
```js
// OPAQUE
material.map.format = THREE.RGBFormat;
material.transparent = false;

// MASK
material.map.format = THREE.RGBAFormat;
// TODO the alpha needs to be used for the mask only
material.alphaTest = alphaCutoff;
material.transparent = false;

// BLEND
material.map.format = THREE.RGBAFormat;
material.transparent = true;
```

**Comment by mrdoob** [comment]
@WestLangley 

> I think the spheres in the test case all have `opaque` true. What is the blending function being used ? No blending? Normal blending? As far as I know, we have not agreed on what the alpha type means for three.js.

As far as I understand... `OPAQUE`, `MASK` and `BLEND` only apply to `material.map`.

> And `diffuseColor.a` can be `1.1`?  What does the compositor do with a value greater than 1? Is it clamped?

As far as I know, values greater than 1 get clamped.

**Comment by WestLangley** [comment]
```js
// OPAQUE
material.map.format = THREE.RGBFormat;
material.transparent = false;
```
I would tend to agree. Of course, that means no alpha test using `map.a`. 

```js
// MASK
material.map.format = THREE.RGBAFormat;
// TODO the alpha needs to be used for the mask only
material.alphaTest = alphaCutoff;
material.transparent = false;
```
Maybe to avoid the transparent restriction it may be sufficient to modify the three.js alpha test code so `alpha > alphaTest` is set to 1.

```js
// BLEND
material.map.format = THREE.RGBAFormat;
material.transparent = true;
```
Yes, I think so.

Maybe we need is a `material.alphaMask` flag? -- or do we just leave this as a `GLTFLoader` thing?

**Comment by WestLangley** [comment]
>As far as I understand... OPAQUE, MASK and BLEND only apply to material.map.

Based on above, they also control the blending function in three.js

**Comment by mrdoob** [comment]
> Maybe to avoid the transparent restriction it may be sufficient to modify the three.js alpha test code so `alpha > alphaTest` is set to 1.

Do you mean that if `alphaTest > 0` we can force diffuse.a to 1.0 after the discard?

**Comment by WestLangley** [comment]
Yes. It is my understand that is what the spec calls for.

**Comment by WestLangley** [comment]
Yes. It is my understanding that is what the spec calls for.

**Pull request merged**